### PR TITLE
Integrate logging into the Blobs Event Processor Store types

### LINKED ISSUE #8663: Integrate logging into the Blobs Event Processor Store types
# Summary

As a library, it is important for the Event Hubs client libraries to provide a means for developers using it to understand its activities and the sequence of operations that comprise them, in order to assist in development and debugging efforts.   The standard .NET event source is used for logging within the core Event Hubs client library, but has not yet been integrated with the Blobs Event Processor Store client library.

# Scope of Work

- Implement logging within the Blobs Event Processor store library types which conforms to the patterns used in the core Event Hubs client library for logging the beginning and completion of key operations, as well as any errors that occurred during them.  

- Operation logging should extend to areas that are likely to be of interest to developers using and troubleshooting event processor store.

- Event source data is encapsulated within its own event source that corresponds to the package name of the library.

- The event source belongs to the `Diagnostics` child namespace within the library, following the conventions used in the core Event Hubs client library.

# Success Criteria

- The necessary types within the Blobs Event Processor Store have integrated logging via the event source, capturing key operations and following the established patterns within the core Event Hubs client library.

- The existing test suite continues to produce deterministic results and pass reliably.

### PR DESCRIPTION
**Pull request by christothes**
closes #8663 

**Test Coverage for** `BlobsCheckpointStore`

Before:
```
+---------+------+--------+--------+
|         | Line | Branch | Method |
+---------+------+--------+--------+
| Total   | 3.9% | 0%     | 9.09%  |
+---------+------+--------+--------+
| Average | 3.9% | 0%     | 9.09%  |
+---------+------+--------+--------+
```

After:
```
+---------+--------+--------+--------+
|         | Line   | Branch | Method |
+---------+--------+--------+--------+
| Total   | 94.93% | 100%   | 100%   |
+---------+--------+--------+--------+
| Average | 94.93% | 100%   | 100%   |
+---------+--------+--------+--------+
```

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/Storage/BlobsCheckpointStore.cs [hunk 1]
```diff
@@ -107,15 +116,17 @@ await foreach (BlobItem blob in ContainerClient.GetBlobsAsync(traits: BlobTraits
                     ));
                 }
 
+                Logger.ListOwnershipAsyncComplete(fullyQualifiedNamespace, eventHubName, consumerGroup, ownershipList.Count);
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/Storage/BlobsCheckpointStore.cs]: We may want to consider wrapping in a `try/finally`  to ensure that the complete event always gets emitted.   I don't have a very strong opinion there, but that's the approach we used for the main Event Hubs library.   You'll always see the "Start" and "Complete" and may or may not see an "Error" between them.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/BlobsCheckpointStore/BlobsCheckpointStoreTests.cs [hunk 2]
```diff
@@ -36,5 +51,301 @@ public void ConstructorRequiresRetryPolicy()
         {
             Assert.That(() => new BlobsCheckpointStore(Mock.Of<BlobContainerClient>(), null), Throws.ArgumentNullException);
         }
+
+        [Test]
+        public async Task ListOwnershipLogsStartAndComplete()
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/BlobsCheckpointStore/BlobsCheckpointStoreTests.cs]: nit:  If you would be so kind, please include doc comments on all members.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/BlobsCheckpointStore/BlobsCheckpointStoreTests.cs [hunk 3]
```diff
@@ -36,5 +51,301 @@ public void ConstructorRequiresRetryPolicy()
...
+        {
+            var blobList = new List<BlobItem>{
+                BlobsModelFactory.BlobItem($"{FullyQualifiedNamespace}/{EventHubName}/{ConsumerGroup}/ownership/{Guid.NewGuid().ToString()}",
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/BlobsCheckpointStore/BlobsCheckpointStoreTests.cs]: You're definitely more thorough than we've been with respect to validating logging.  I'm a bit on the fence; while I think that its great, I'm concerned about potentially setting a bar for rigor around logging that we may not want to apply consistently.

For now, I think I'm content to leave things as you've written.   I'll think more on whether we should be doing the same across other areas.

**Review by jsquire — COMMENTED** [review]
Overall, this looks great.   I made a few requests around formatting to help align with the rest of the Event Hubs areas.

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/BlobsCheckpointStore/BlobsCheckpointStoreTests.cs (hunk 3 above)
Comment (christothes) [reply sdk/eventhub/Azure.Messaging.EventHubs.Processor/tests/BlobsCheckpointStore/BlobsCheckpointStoreTests.cs]: I've been bitten by too many event source logging bugs in my time, perhaps :)

File: sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/Storage/BlobsCheckpointStore.cs [hunk 4]
```diff
@@ -91,44 +91,49 @@ internal sealed class BlobsCheckpointStore : PartitionManager
...
             cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            List<PartitionOwnership> result = null;
 
-            var prefix = string.Format(OwnershipPrefix, fullyQualifiedNamespace.ToLower(), eventHubName.ToLower(), consumerGroup.ToLower());
-
-            Func<CancellationToken, Task<IEnumerable<PartitionOwnership>>> listOwnershipAsync = async listOwnershipToken =>
+            try
             {
-                var ownershipList = new List<PartitionOwnership>();
+                var prefix = string.Format(OwnershipPrefix, fullyQualifiedNamespace.ToLower(), eventHubName.ToLower(), consumerGroup.ToLower());
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/Storage/BlobsCheckpointStore.cs]: Should we consider `ToLowerInvariant()` here?

**Review by jsquire — APPROVED** [review]
LGTM.   I left one (very) minor thought, which is definitely non-blocking.    

@christothes: I'm happy to merge with a final thumbs up that you're ready.

**Comment by christothes** [comment]
@jsquire I’m good with the merge. Thanks!

**Pull request merged**